tensorwaves/estimator.py

```diff
@@ -1,11 +1,49 @@
+"""Defines estimators which estimate a model's ability to represent the data.
+
+All estimators have to implement the `~.interfaces.Estimator` interface.
+"""
+import numpy as np
+
 import tensorflow as tf
 
+from tensorwaves.interfaces import Estimator, Function
+
+
+class UnbinnedNLL(Estimator):
+    """Unbinned negative log likelihood estimator.
+
+    Args:
+        model: A model that should be compared to the dataset.
+        dataset: The dataset used for the comparison. The model has to be
+            evaluateable with this dataset.
+    """
+
+    def __init__(self, model: Function, dataset: dict) -> None:
+        self.__model = model
+        self.__dataset = dataset
+
+    def __call__(self) -> float:
+        props = self.__model(self.__dataset)
+        logs = tf.math.log(props)
+        log_lh = tf.reduce_sum(logs)
+        return -log_lh.numpy()
+
+    def gradient(self) -> np.ndarray:
+        # for i, p in enumerate(float_pars):
+        #     p.update(par[i])
+        # grad.n += 1
+        # with tf.GradientTape() as gradient:
+        #     gradient.watch(float_pars)
+        #     nll_val = nll(*args)
+        # g = gradient.gradient(
+        #     nll_val, float_pars, unconnected_gradients=tf.UnconnectedGradients.ZERO)
+        # g_val = [i.numpy() for i in g]
+        # return g_val
+        pass
 
-class UnbinnedLL(tf.keras.losses.Loss):
-    def __init__(self):
-        super(UnbinnedLL, self).__init__(name="UnbinnedLL")
+    @property
+    def parameters(self) -> dict:
+        return self.__model.parameters
 
-    def call(self, y_true, y_pred):
-        logs = tf.math.log(y_pred)
-        ll = tf.reduce_sum(logs)
-        return ll
+    def update_parameters(self, new_parameters: dict) -> None:
```

Comment:
Why not a setter?

Reply:
To me the `update_parameters` makes it clearer that the you are not replacing parameters but merely updating some of their values